### PR TITLE
add sample_weight support for match models

### PR DESCRIPTION
**Pull request by eric-gecheng**

### CODE REVIEW

File: tzrec/models/dbmtl.py [hunk 1]
```diff
@@ -31,12 +31,13 @@ class DBMTL(MultiTaskRank):
         model_config (ModelConfig): an instance of ModelConfig.
         features (list): list of features.
         labels (list): list of label names.
+        sample_weights (list): sample weight names.
     """
 
     def __init__(
-        self, model_config: ModelConfig, features: List[BaseFeature], labels: List[str]
+        self, model_config: ModelConfig, features: List[BaseFeature], labels: List[str], sample_weights: List[str] = None
```
Comment (tiankongdeguiji) [inline tzrec/models/dbmtl.py]: - add **kwargs to all models for better compatibility
- sample_weights typing is Optional[List[str]]

File: tzrec/models/match_model.py [hunk 2]
```diff
@@ -188,7 +190,7 @@ def _init_loss_impl(self, loss_cfg: LossConfig, suffix: str = "") -> None:
         assert (
             loss_type == "softmax_cross_entropy"
         ), "match model only support softmax_cross_entropy loss now."
```
Comment (tiankongdeguiji) [inline tzrec/models/match_model.py:197]:
```python
reduction = 'none' if self._sample_weight else 'mean'
self._loss_modules[loss_name] = nn.CrossEntropyLoss(reduction=reduction)
```

File: tzrec/models/match_model.py [hunk 3]
```diff
@@ -220,7 +223,7 @@ def _loss_impl(
             label = _arange_int_label(pred)
         else:
             label = _zero_int_label(pred)
-        losses[loss_name] = self._loss_modules[loss_name](pred, label)
+        losses[loss_name] = torch.mean(self._loss_modules[loss_name](pred, label) * sample_weight)
```
Comment (tiankongdeguiji) [inline tzrec/models/match_model.py]:
```python
losses[loss_name] = self._loss_modules[loss_name](pred, label)
if self._sample_weight:
    losses[loss_name] = torch.mean(losses[loss_name] * sample_weight)
```

File: tzrec/models/rank_model.py [hunk 4]
```diff
@@ -42,15 +42,17 @@ class RankModel(BaseModel):
...
     def __init__(
         self,
         model_config: model_pb2.ModelConfig,
         features: List[BaseFeature],
         labels: List[str],
+        sample_weights: List[str] = None
     ) -> None:
-        super().__init__(model_config, features, labels)
+        super().__init__(model_config, features, labels, sample_weights)
         self._num_class = model_config.num_class
```
Comment (tiankongdeguiji) [inline tzrec/models/rank_model.py:57]: add sample weight for loss of rank and multi_task_rank models later?

**Review by tiankongdeguiji — CHANGES_REQUESTED** [review]

File: tzrec/models/dbmtl.py (hunk 1 above)
Comment (eric-gecheng) [reply tzrec/models/dbmtl.py]: resolved

File: tzrec/models/match_model.py (hunk 2 above)
Comment (eric-gecheng) [reply tzrec/models/match_model.py:197]: resolved

File: tzrec/models/match_model.py (hunk 3 above)
Comment (eric-gecheng) [reply tzrec/models/match_model.py]: resolved

File: tzrec/models/rank_model.py (hunk 4 above)
Comment (eric-gecheng) [reply tzrec/models/rank_model.py:57]: rank models will be fixed in the next PR.

**Review by tiankongdeguiji — APPROVED** [review]
LGTM
